src/hypothesis/extra/datetime.py

```diff
@@ -23,62 +32,185 @@
+            else:
+                raise InvalidArgument((
+                    'min_datetime=%s and max_datetime=%s must not mix naive '
+                    'and tz-aware values.  See the documentation for how each '
+                    'of these modes work.') % (self.min_dt, self.max_dt))
+        if allow_naive and self.min_dt.tzinfo is not None:
+            raise InvalidArgument(
+                'Cannot create naive datetimes between tz-aware bounds as '
+                'they are not comparable; do not set allow_naive=True.')
+        check_valid_interval(
+            self.min_dt, self.max_dt, 'min_datetime', 'max_datetime')
 
     def do_draw(self, data):
         while True:
             try:
-                result = dt.datetime(
-                    year=cu.centered_integer_range(
-                        data, self.min_year, self.max_year, 2000
-                    ),
-                    month=cu.integer_range(data, 1, 12),
-                    day=cu.integer_range(data, 1, 31),
-                    hour=cu.integer_range(data, 0, 24),
-                    minute=cu.integer_range(data, 0, 59),
-                    second=cu.integer_range(data, 0, 59),
-                    microsecond=cu.integer_range(data, 0, 999999)
-                )
-                if (
-                    not self.allow_naive or
-                    (self.timezones and cu.boolean(data))
-                ):
-                    result = cu.choice(data, self.timezones).localize(result)
-                return result
-
+                # Constuct a datetime within bounds, ignoring tzinfo for now
+                result = dt.datetime.min
+                constrained = True
+                for name in ('year', 'month', 'day',
+                             'hour', 'minute', 'second', 'microsecond'):
+                    low = getattr(dt.datetime.min, name)
+                    high = getattr(dt.datetime.max, name)
+                    if constrained:
+                        low = max(low, getattr(self.min_dt, name))
+                        high = min([high, getattr(self.max_dt, name)])
+                    if name == 'year':
+                        val = cu.centered_integer_range(data, low, high, 2000)
+                    else:
+                        val = cu.integer_range(data, low, high)
+                    result = result.replace(**{name: val})
+                    constrained = constrained and val in (low, high)
+                # Naive bounds: just localise to the timezone (if any)
+                if not self.timezones or self.allow_naive and cu.boolean(data):
+                    return result
+                to_tz = cu.choice(data, self.timezones)
+                if self.min_dt.tzinfo is None:
+                    return to_tz.normalize(to_tz.localize(result))
+                # Aware bounds: localise to UTC, then convert properly
+                return to_tz.normalize(
+                    pytz.utc.localize(result).astimezone(to_tz))
```

Comment:
This will require a minor redesign so we can still handle bounds with different timezones without biasing the drawn values, but you're right.

Correct logic is to not convert bounds to UTC, then draw for which of the bounds-timezones to localise in and then in a try-catch loop draw a timezone and convert to it.  This also means that the timedelta between aware bounds may be greater than that between naive bounds.